### DISCUSSION TITLE
Add support for outputting result of extracted package tree and pending updates

### DISCUSSION THREAD

**Discussion by jamietanna** [renovatebot/renovate#24823] [Suggest an Idea]
### What would you like Renovate to be able to do?

Discussed in #13150, and being used by a few folks, we should make it possible for (self-hosted) Renovate to output the fully extracted dependency tree, which may also include a list of pending updates depending on how Renovate is being run.

<details>
<summary>Previous discussion</summary>

> As blogged about in https://www.jvt.me/posts/2022/11/01/renovate-dependency-graph/ I've pulled this out with Typescript through https://gitlab.com/tanna.dev/renovate-graph - I'll also raise a feature request to get this officially supported 🤞🏽
> This is used as part of further tooling I've built which can be found linked out from https://dmd.tanna.dev, and includes a command-line tool `dmd` for better querying of the data

_Originally posted by @jamietanna in https://github.com/renovatebot/renovate/discussions/13150#discussioncomment-4029477_

</details>

This appears to be something that other folks would like (#13150) as well as being used as part of https://gitlab.com/tanna.dev/renovate-graph and https://dmd.tanna.dev

### If you have any ideas on how this should be implemented, please tell us here.

This can be picked up from the [`ExtractResult`](https://github.com/renovatebot/renovate/blob/1767f7608b10b3100c7dd5f0929ba4eb4ce18a2f/lib/workers/types.ts#L191-L194) and then output to disk.

This would ideally write to disk with a name such as:

- `${platform}-${organisation}-${repo}` if appropriate

Longer term, this could also write to S3, but a good start would be to disk.

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by rarkins** [comment]
This type of feature has the potential to be a problem, like "gifting" a puppy to someone is not really a gift, it's a burden.

Today we can output JSON-formatted logs, and can output them to disk too. I would prefer that it's used instead.

**Comment by jamietanna** [comment]
Interesting, that's fair - so I guess the preference would be for folks to keep to `env LOG_LEVEL=debug LOG_FORMAT=json renovate ...` and then parse those log lines as they appear?

**Comment by rarkins** [comment]
If you configure `logFile` then it automatically outputs in JSON format and debug log level

**Comment by jamietanna** [comment]
Interesting, thanks! That looks like that may do it for what we're looking at (trying to get the same data we use `renovate-graph` for, but without duplicating as much of the always-running aspect of Renovate On-Prem, or trying to run `renovate-graph` with update lookups as well as already doing that with Renovate)

[Related ticket in DMD](https://gitlab.com/tanna.dev/dependency-management-data/-/issues/65)

**Comment by jamietanna** [comment]
FYI we've had some investigation with using Renovate's debug logs for this, but are not sure this will work for our use case.

I think in this specific case - where we want a view of all the dependencies that a given repo has - we'd need to run Renovate with `requireConfig=ignored`, which wouldn't work well with using Renovate for updates.

It may be we'll run one Renovate for dependency updates and one for purely retrieving dependency data, but not decided yet.

**Comment by jdeastwood** [comment]
It would be very cool if this could support outputting the graph to the [Github dependency submission API](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/using-the-dependency-submission-api#using-pre-made-actions). Then [ecosystems that aren't already automatically scraped](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/about-the-dependency-graph#supported-package-ecosystems) by Github could benefit from a low-config way to use the github dependency graph. May also be useful in other git-hosting ecosystems!

**Reply by jamietanna** [reply]
@jdeastwood I've just been giving it a play, using a tool I've written, `renovate-to-sbom` ([blog post](https://www.jvt.me/posts/2023/11/03/renovate-to-sbom/)) and unfortunately GitHub's API is rather restrictive.

For instance, using https://github.com/dagger/dagger, we detect [some Elixir dependencies](https://github.com/dagger/dagger/blob/4020381db19404686eaf158e1f8b76fcc57a7e05/sdk/elixir/mix.exs#L26-L35).

When submitting these to the API, GitHub responds with:

```json
{
  "message": "invalid package url: in manifest \"github/dagger/dagger\" decoding \"pkg:mix/nestru@~%3E%200.3\": invalid package url type: mix",
  "documentation_url": "https://docs.github.com/rest/dependency-graph/dependency-submission#create-a-snapshot-of-dependencies-for-a-repository"
}
```

It looks like they're only expecting "known" package types, whereas the libraries I've used to parse package-urls are much more forgiving when there's an unknown type :disappointed:

**Reply by jamietanna** [reply]
I'm asking clarification in https://github.com/package-url/purl-spec/issues/286 to see if it _is_ expected by GitHub, and I've raised with GitHub too https://github.com/orgs/community/discussions/87501

**Reply by jamietanna** [reply]
@jdeastwood will hopefully have some more shared shortly, it looks like GitHub will only support a subset of package URL (pURLs) so we need to make more effort in `renovate-to-sbom` to 

But other than that, it should Just Work :tm: and in fact I've got an example repo where I'm picking up Elixir dependencies, which GitHub doesn't officially support - https://github.com/jamietanna/dagger/network/dependencies?q=absinthe_client

**Reply by jamietanna** [reply]
@jdeastwood all working :eyes: :rocket:  https://www.jvt.me/posts/2024/01/16/renovate-github-dependency-submission/

(doesn't support _all_ of the ecosystems that Renovate does, but it's more than GitHub does out-of-the-box!)

**Reply by jdeastwood** [reply]
Thanks for the heads up @jamietanna! I'll check it out!

**Comment by jamietanna** [comment]
I've just finalised [support for parsing debug logs](https://gitlab.com/tanna.dev/dependency-management-data/-/issues/85) this within [dependency-management-data](https://dmd.tanna.dev) and a key question I've got is whether we've got anything in the codebase to catch an inadvertent change to the log message `packageFiles with updates` as a breaking change?

I.e. there appear to be several folks relying on the functionality ([via](https://github.com/renovatebot/renovate/discussions/13150)) it may be useful to know ahead-of-time if it's going to break anyone relying on it.

**Reply by rarkins** [reply]
No. We don't plan to make this a "contract" with external parties or tools

**Comment by phyzical** [comment]
@jamietanna just wanted to say thanks, given my testing repo your linked workflow pretty much just works out of the box

if anyones after a quick ref in github to it all 

If your project is public this workflow should just work

https://github.com/phyzical/renovate-tester/blob/main/.github/workflows/deps.yml
https://github.com/phyzical/renovate-tester/network/dependencies?page=2
